Fix GetFunctionEnter3Info on linux

davmason opened the pull request:
Fixes #19622

When slow path ELT is enabled on linux, this will allocate 128 bytes of stack buffer for each function that the profiler can use to copy any enregistered structs to. I chose 128 because that is the size of the available general purpose registers + the available floating point registers for argument passing (since we store the xmm values in the profiler code as 64 bit). Previously we were allocatng 216 bytes of stack space to store the registers, so while this grows the stack usage by ~50% I believe this is an acceptable tradeoff for slow path ELT.

`ProfileArgIterator::CopyStructFromRegisters` is `ArgLocDesc::CopyStructToRegisters` but modified to work with `PROFILE_PLATFORM_SPECIFIC_DATA` instead of `TransitionBlock`, and also to copy out of registers instead of in to registers. `PROFILE_PLATFORM_SPECIFIC_DATA` has a different layout, and uses `UINT64` for all float registers, including the 128 bit registers.

I tested with this program, everything in the foo* functions behaved as I expected. If there are cases I missed please let me know.

```
using System;

namespace linux_elt
{
    struct SampleStruct
    {
        public int a;
        public double b;
        public int c;
        public float d;
    }

    class Program
    { 
        static int foo(int a, int b, int c, string d)
        {
            return a + b + c + d.Length;
        }

        static void foo2(int a, int b, int c, int d, int e, int f, int g, int h,
                        int i, int j, int k, int l, int m, int n, int o, int p,
                        int q, int r, int s, int t, int u, int v, int w, int x,
                        int y, int z)
        {
            int sum = a + b + c + d + e + f + g + h + i + j + k + l + m + n;
            Console.WriteLine($"Sum={sum}");
        }

        static void foo3(SampleStruct ss, float f)
        {
            Console.WriteLine($"foo3={(int)(ss.a + ss.b + ss.c + ss.d + f)}");
        }

        static void foo4(int d, ref int a, out int b, int c)
        {
            b = 10;
            Console.WriteLine($"foo4={a + b + c + d}");
        }

        static void Main(string[] args)
        {
            Console.WriteLine("*****Calling foo!*****");
            Console.WriteLine("foo: {0:d}", foo(0xaa, 0xbb, 0xcc, "bar"));
            foo2(1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18,
                 19, 20, 21, 22, 23, 24, 25, 26);

            SampleStruct ss = new SampleStruct();
            ss.a = 1;
            ss.b = 2.0;
            ss.c = 3;
            ss.d = 4.0f;
            foo3(ss, 3.72f);

            int a = 5;
            int b;
            foo4(7, ref a, out b, 13);
        }
    }
}
```

davmason commented on src/vm/amd64/profiler.cpp:
```diff
@@ -204,6 +215,85 @@ ProfileArgIterator::~ProfileArgIterator()
+
+    LPVOID destOrig = (LPVOID)&pData->buffer[pData->bufferPos];
+    //Increase bufferPos by ceiling(fieldBytes/8)
+    pData->bufferPos += (fieldBytes + 7) / 8;
```
I need to add an asset that checks that the buffer never overflows

janvorli commented on src/vm/amd64/profiler.cpp:
```diff
@@ -204,6 +215,85 @@ ProfileArgIterator::~ProfileArgIterator()
+    EEClass* eeClass = argLocDesc->m_eeClass;
+    _ASSERTE(eeClass != NULL);
+
+    // We start at the first eightByte that the destOffset didn't skip completely.
```
There is no destOffset in this copy of the code, please remove this comment.

janvorli commented on src/vm/proftoeeinterfaceimpl.h:
```diff
@@ -71,6 +71,10 @@ class ProfileArgIterator
     }
+    
+#ifdef UNIX_AMD64_ABI
+    LPVOID ProfileArgIterator::CopyStructFromRegisters();
```
Please remove the class name here.

mjsabby commented:
@davmason there is a sample clr profiler that tried to do ELT and failed in the clr-samples repo if you'd like to validate that.

davmason commented on src/vm/amd64/asmhelpers.S:
```diff
@@ -39,8 +39,10 @@
 //
 .equ SIZEOF_PROFILE_PLATFORM_SPECIFIC_DATA, 0x8*22 + 0x8  // includes fudge to make FP_SPILL right
```
Yeah, that conceptually makes more sense

davmason commented on src/vm/amd64/asmhelpers.S:
```diff
@@ -39,8 +39,10 @@
+.equ SIZEOF_PROFILE_STRUCT_BUFFER, 0x8*16
 
-.equ SIZEOF_STACK_FRAME, SIZEOF_PROFILE_PLATFORM_SPECIFIC_DATA + SIZEOF_FP_ARG_SPILL
+.equ SIZEOF_STACK_FRAME, SIZEOF_PROFILE_PLATFORM_SPECIFIC_DATA + SIZEOF_FP_ARG_SPILL + SIZEOF_PROFILE_STRUCT_BUFFER
```
Doh! I was thinking about as if it were a member on ProfileArgIterator. I changed it to actually be a member, I think it fits better there since it is just an implementation detail on creating the PROFILE_PLATFORM_SPECIFIC_DATA

davmason commented:
Thanks @noahfalk and @janvorli! I addressed your feedback, and am going to re-run the tests I have plus the profiler Mukul pointed out, then merge once everything is green.

davmason commented:
@mjsabby what failures were you seeing in the sample profiler? Was it crashing or giving wrong output? I Just ran both from https://github.com/microsoft/clr-samples/tree/master/ProfilingAPI and they both execute without crashing and seem to give coherent output.

davmason commented:
> 
> 
> @mjsabby what failures were you seeing in the sample profiler? Was it crashing or giving wrong output? I Just ran both from https://github.com/microsoft/clr-samples/tree/master/ProfilingAPI and they both execute without crashing and seem to give coherent output.

Nevermind, I see in https://github.com/microsoft/clr-samples/issues/12 that segfaults were the issue. I was seeing segfaults before my change that are solved now, so I believe this change fixes that issue

mjsabby commented:
Thanks for confirming @davmason. Now that you have fixed this for Linux it's not as urgent (i.e. in 3.0 timeframe) for me to have an issue where I'd like to propose that we innovate in this area and provide argument and return value inspection for fast path ELT.
